File: packages/deon-python/source/deon/jsonread.py
"""JSON, as Deon values.

The mapping is specification 9.1, and one line of it decides the whole design:

> JSON number to its source token spelling.

Not to its value — to *the way it was written*. `1.50` is the string `"1.50"`, not `"1.5"`, and `1e3`
is `"1e3"`, not `"1000"`. A Deon value is a string, so a number that arrived through a `float` and
came back out through a `repr` would already have lost the only thing about it that was ever there.

Python's `json` can do this, which is the one place Python is kinder than a language without a
batteries-included decoder: `parse_float` and `parse_int` are handed the **raw matched substring**, so
asking for `str` preserves the spelling exactly. But the stdlib decoder must be told three further
things, and each of them is silent when it is wrong:

1. `parse_constant` must be supplied *and must raise*. By default `json.loads('[NaN, Infinity]')`
   succeeds and yields floats. Specification 9.1 enumerates what a JSON value may be, and those are
   not in it.
2. Repeated members need `object_pairs_hook`. The default decoder builds a `dict`, which keeps a
   repeated key in the slot it *first* appeared in — and specification 9.1 defers to Deon's
   last-write-wins rule, which moves the key to its *final* write position (specification 5).
3. `bool` is an `int` in Python. `True` must be tested for with `is`, before any numeric test, or a
   boolean quietly becomes a number.
"""

from __future__ import annotations

import json

from .diagnostic import DiagnosticCode, Span, error
from .parser import MAX_DEPTH
from .value import DeonMap, Value


class _Refused(Exception):
    pass


def _refuse_constant(literal: str):
    raise _Refused(f"'{literal}' is not a JSON value Deon admits.")


def _pairs(pairs) -> DeonMap:
    deon_map = DeonMap()

    for key, value in pairs:
        # `insert`, not assignment: a repeated member moves the key to its final write position.
        deon_map.insert(key, value)

    return deon_map


def _convert(node, depth: int, span: Span) -> Value:
    if depth > MAX_DEPTH:
        raise error(
            DiagnosticCode.RESOURCE_FORMAT,
            "The imported resource nests more deeply than Deon will read.",
            span,
        )

    # `is`, and before everything: `True == 1` and `isinstance(True, int)` are both true in Python,
    # so a boolean tested for numerically is a boolean silently read as a number.
    if node is True:
        return "true"

    if node is False:
        return "false"

    if node is None:
        return ""

    if isinstance(node, str):
        return node

    if isinstance(node, list):
        return [_convert(item, depth + 1, span) for item in node]

    if isinstance(node, DeonMap):
        return DeonMap([(key, _convert(value, depth + 1, span)) for key, value in node.items()])

    # Nothing else can arrive: numbers were handed to `str` by the decoder, and the constants were
    # refused. If something does, it is a bug here rather than a fact about the document.
    raise error(
        DiagnosticCode.RESOURCE_FORMAT,
        f"A JSON {type(node).__name__} is not a value Deon can read.",
        span,
    )


def read_json(text: str, span: Span) -> Value:
    """JSON text, as a Deon value.

    `span` is where the failure will be reported, which for an imported resource is the statement that
    imported it rather than anywhere inside the resource itself.
    """
    try:
        decoded = json.loads(
            text,
            parse_float=str,
            parse_int=str,
            parse_constant=_refuse_constant,
            object_pairs_hook=_pairs,
        )
    except _Refused as refusal:
        raise error(DiagnosticCode.RESOURCE_FORMAT, str(refusal), span) from None
    except RecursionError:
        raise error(
            DiagnosticCode.RESOURCE_FORMAT,
            "The imported resource nests more deeply than Deon will read.",
            span,
        ) from None
    except ValueError as failure:
        raise error(DiagnosticCode.RESOURCE_FORMAT, f"Invalid JSON: {failure}.", span) from None

    return _convert(decoded, 1, span)

File: packages/deon-go/jsonread.go
package deon

import (
	"encoding/json"
	"strings"
)

// jsonToValue converts JSON to a Deon value (specification 9.1). The one rule that shapes everything
// is that a number keeps its *source token spelling*: 1.50 converts to the string "1.50" and not to
// "1.5". Go's encoding/json preserves that when the decoder is told to hand numbers back as
// json.Number, which is the literal text — so the standard library reader is enough, and no
// hand-written number parser is needed.
//
// A boolean becomes "true" or "false", null becomes the empty string, and a repeated object member
// follows Deon's last-write-wins map rule, which is why an ordered Map is built by hand from the token
// stream rather than decoding into a Go map that would lose both the order and the move.
func jsonToValue(data string, at Span) Value {
	decoder := json.NewDecoder(strings.NewReader(data))
	decoder.UseNumber()

	value, err := readJSONValue(decoder, at)
	if err != nil {
		fail(ResourceFormat, "The imported JSON is invalid: "+err.Error(), at)
	}

	// Trailing content after one JSON value is malformed input.
	if decoder.More() {
		fail(ResourceFormat, "The imported JSON has trailing content.", at)
	}

	return value
}

func readJSONValue(decoder *json.Decoder, at Span) (Value, error) {
	token, err := decoder.Token()
	if err != nil {
		return nil, err
	}
	return readJSONFrom(token, decoder, at)
}

func readJSONFrom(token json.Token, decoder *json.Decoder, at Span) (Value, error) {
	switch t := token.(type) {
	case json.Delim:
		switch t {
		case '{':
			return readJSONObject(decoder, at)
		case '[':
			return readJSONArray(decoder, at)
		default:
			fail(ResourceFormat, "The imported JSON is malformed.", at)
		}
	case string:
		return t, nil
	case json.Number:
		// The literal spelling, exactly as written.
		return t.String(), nil
	case bool:
		if t {
			return "true", nil
		}
		return "false", nil
	case nil:
		return "", nil
	}
	fail(ResourceFormat, "The imported JSON holds a value Deon cannot represent.", at)
	return nil, nil
}

func readJSONObject(decoder *json.Decoder, at Span) (Value, error) {
	result := NewMap()
	for decoder.More() {
		keyToken, err := decoder.Token()
		if err != nil {
			return nil, err
		}
		key, ok := keyToken.(string)
		if !ok {
			fail(ResourceFormat, "A JSON object key must be a string.", at)
		}
		value, err := readJSONValue(decoder, at)
		if err != nil {
			return nil, err
		}
		result.Set(key, value)
	}
	// Consume the closing '}'.
	if _, err := decoder.Token(); err != nil {
		return nil, err
	}
	return result, nil
}

func readJSONArray(decoder *json.Decoder, at Span) (Value, error) {
	result := []Value{}
	for decoder.More() {
		value, err := readJSONValue(decoder, at)
		if err != nil {
			return nil, err
		}
		result = append(result, value)
	}
	// Consume the closing ']'.
	if _, err := decoder.Token(); err != nil {
		return nil, err
	}
	return result, nil
}
